driver-core/src/main/com/mongodb/internal/connection/AwsAuthenticator.java

```diff
@@ -77,12 +83,24 @@ protected SaslClient createSaslClient(final ServerAddress serverAddress) {
 
     private static class AwsSaslClient implements SaslClient {
         private final MongoCredential credential;
+        @Nullable
+        private final AwsCredential credentialFromSupplier;
         private final byte[] clientNonce = new byte[RANDOM_LENGTH];
         private int step = -1;
         private String httpResponse;
 
         AwsSaslClient(final MongoCredential credential) {
```

Comment:
For the new mechanism based on `Supplier<AwsCredential>` to work, a new `AwsSaslClient` must be created each time authentication is being done. However, it seems that at least `ScramShaAuthenticator` stores it in its field `speculativeSaslClient` and reuses in `createSpeculativeAuthenticateCommand`. I may misunderstand the situation.

Reply:
It's not strictly required that we get new credentials for every single authentication attempt.  These are not one-time passwords.  So even if this is true, it's a short-lived situation that only lasts the length of a single handshake.

Reply:
> it's a short-lived situation that only lasts the length of a single handshake

👍